### ISSUE TITLE
AWS: First salt-run state.orchestrate butler.deploy on the salt-master fails

### ISSUE THREAD

**Issue by brainstorm** [llevar/butler#18]
Hello Sergei,

I'm giving a try to the AWS deployment example:

https://github.com/llevar/butler/tree/master/examples/deployment/aws/large-cluster

Deploying on AWS with terraform went beautifully. I think that one missing bit there is letting the reader know on how to configure a convenience bastion host SSH configuration, like so:

```
Host butler-jump
	HostName 13.54.105.9
	User centos

Host butler
	HostName 10.0.0.59
	User centos
	ProxyCommand ssh butler-jump -W %h:%p
```

Other than this minor documentation detail, following the docs I'm reaching the following point/errors when running salt-run on the saltmaster as root as instructed:

```
[root@salt-master centos]# salt-run state.orchestrate butler.deploy
No minions matched the target. No command was sent, no jid was assigned.
[ERROR   ] No minions returned

[ERROR   ] Run failed on minions: salt-master
Failures:
    salt-master:
      Name: unzip - Function: pkg.installed - Result: Clean Started: - 20:41:03.292658 Duration: 799.128 ms
      Name: /opt/consul/agent - Function: archive.extracted - Result: Clean Started: - 20:41:04.092709 Duration: 681.057 ms
      Name: /opt/consul/agent/consul - Function: file.managed - Result: Clean Started: - 20:41:04.773951 Duration: 1.322 ms
      Name: /usr/bin/consul - Function: file.symlink - Result: Clean Started: - 20:41:04.776017 Duration: 0.891 ms
      Name: /var/consul - Function: file.directory - Result: Clean Started: - 20:41:04.777043 Duration: 0.784 ms
      Name: /etc/opt/consul.d - Function: file.directory - Result: Clean Started: - 20:41:04.777940 Duration: 0.794 ms
      Name: /etc/opt/consul.d/ - Function: file.recurse - Result: Clean Started: - 20:41:04.778857 Duration: 146.589 ms
      Name: dnsmasq - Function: pkg.installed - Result: Clean Started: - 20:41:04.925585 Duration: 0.533 ms
      Name: /etc/dnsmasq.conf - Function: file.managed - Result: Changed Started: - 20:41:04.928108 Duration: 148.703 ms
      Name: /etc/dnsmasq.d/10-consul - Function: file.append - Result: Changed Started: - 20:41:05.077020 Duration: 2.663 ms
      Name: dnsmasq - Function: service.running - Result: Changed Started: - 20:41:05.080467 Duration: 106.88 ms
      Name: /etc/resolv.conf - Function: file.prepend - Result: Changed Started: - 20:41:05.187554 Duration: 2.751 ms
      Name: collectd - Function: pkg.installed - Result: Clean Started: - 20:41:05.190422 Duration: 0.507 ms
      Name: collectd-postgresql.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:41:05.191059 Duration: 50408.015 ms
      Name: collectd-java.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:41:55.711575 Duration: 31318.094 ms
      Name: /lib64/libjvm.so - Function: file.symlink - Result: Changed Started: - 20:42:27.112308 Duration: 59.09 ms
      Name: collectd-rrdtool.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:42:27.197114 Duration: 26954.658 ms
      Name: collectd-generic-jmx.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:42:54.241479 Duration: 17436.368 ms
      Name: collectd-write_sensu.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:43:11.712289 Duration: 17791.937 ms
      Name: /etc/collectd.conf - Function: file.managed - Result: Changed Started: - 20:43:30.168852 Duration: 266.859 ms
      Name: /usr/share/collectd/types.d - Function: file.managed - Result: Changed Started: - 20:43:30.436396 Duration: 115.128 ms
      Name: collectd - Function: service.running - Result: Changed Started: - 20:43:30.552648 Duration: 520.615 ms
      Name: /var/log/collectd.log - Function: file.managed - Result: Changed Started: - 20:43:31.073546 Duration: 178.152 ms
      Name: install_filebeat - Function: pkg.installed - Result: Changed Started: - 20:43:31.257289 Duration: 5533.408 ms
      Name: /etc/filebeat/filebeat.yml - Function: file.managed - Result: Changed Started: - 20:43:36.793821 Duration: 243.455 ms
      Name: filebeat - Function: service.enabled - Result: Changed Started: - 20:43:37.527077 Duration: 202.969 ms
      Name: filebeat - Function: service.running - Result: Changed Started: - 20:43:37.730937 Duration: 69.706 ms
      Name: libpcap - Function: pkg.installed - Result: Clean Started: - 20:43:37.805809 Duration: 0.473 ms
      Name: install_packetbeat - Function: pkg.installed - Result: Changed Started: - 20:43:37.806381 Duration: 5844.761 ms
      Name: /etc/packetbeat/packetbeat.yml - Function: file.managed - Result: Changed Started: - 20:43:43.677520 Duration: 218.495 ms
      Name: packetbeat - Function: service.enabled - Result: Changed Started: - 20:43:44.387993 Duration: 160.621 ms
      Name: packetbeat - Function: service.running - Result: Changed Started: - 20:43:44.548811 Duration: 87.587 ms
      Name: ntp - Function: pkg.installed - Result: Changed Started: - 20:43:44.641448 Duration: 21328.646 ms
      Name: ntpd - Function: service.enabled - Result: Changed Started: - 20:44:07.170765 Duration: 109.051 ms
      Name: ntpd - Function: service.running - Result: Changed Started: - 20:44:07.280036 Duration: 174.655 ms
      Name: /etc/opt/consul.d/server-config.json - Function: file.managed - Result: Clean Started: - 20:44:07.494604 Duration: 139.291 ms
      Name: /usr/lib/systemd/system/consul-server.service - Function: file.managed - Result: Clean Started: - 20:44:07.634074 Duration: 56.638 ms
      Name: /etc/opt/consul.d/influxdb_consul.json - Function: file.managed - Result: Changed Started: - 20:44:07.691808 Duration: 119.135 ms
      Name: /etc/opt/consul.d/grafana_consul.json - Function: file.managed - Result: Changed Started: - 20:44:07.811161 Duration: 158.401 ms
      Name: consul-server - Function: service.running - Result: Changed Started: - 20:44:08.008765 Duration: 2059.765 ms
    ----------
              ID: join-all-consul-members
        Function: cmd.run
            Name: consul join 10.0.0.59
          Result: False
         Comment: Command "consul join 10.0.0.59" run
         Started: 20:44:10.221437
        Duration: 69.428 ms
         Changes:
                  ----------
                  pid:
                      14625
                  retcode:
                      1
                  stderr:
                      Error joining address '10.0.0.59': Put http://127.0.0.1:8500/v1/agent/join/10.0.0.59: dial tcp 127.0.0.1:8500: getsockopt: connection refused
                      Failed to join any nodes.
                  stdout:
      Name: influxdb - Function: pkg.installed - Result: Changed Started: - 20:44:10.327802 Duration: 31209.716 ms
      Name: influxdb - Function: user.present - Result: Changed Started: - 20:44:42.446100 Duration: 282.127 ms
      Name: /etc/influxdb/influxdb.conf - Function: file.managed - Result: Changed Started: - 20:44:42.730048 Duration: 147.816 ms
      Name: influxdb - Function: service.running - Result: Changed Started: - 20:44:42.879108 Duration: 84.587 ms
      Name: /var/lib/.influxdb - Function: file.directory - Result: Changed Started: - 20:44:42.963972 Duration: 1.546 ms
      Name: /var/lib/.influxdb/data - Function: file.directory - Result: Changed Started: - 20:44:42.965636 Duration: 1.256 ms
      Name: /var/lib/.influxdb/meta - Function: file.directory - Result: Changed Started: - 20:44:42.967029 Duration: 1.298 ms
      Name: /var/lib/.influxdb/hh - Function: file.directory - Result: Changed Started: - 20:44:42.968449 Duration: 1.251 ms
      Name: /var/lib/.influxdb/wal - Function: file.directory - Result: Changed Started: - 20:44:42.969842 Duration: 1.283 ms
      Name: initscripts - Function: pkg.installed - Result: Clean Started: - 20:44:42.988186 Duration: 0.517 ms
      Name: fontconfig - Function: pkg.installed - Result: Clean Started: - 20:44:42.988815 Duration: 0.467 ms
      Name: grafana - Function: pkgrepo.managed - Result: Changed Started: - 20:44:43.006339 Duration: 519.696 ms
      Name: grafana - Function: pkg.installed - Result: Changed Started: - 20:44:43.526208 Duration: 71889.838 ms
      Name: /etc/grafana/grafana.ini - Function: file.managed - Result: Changed Started: - 20:45:56.742311 Duration: 144.832 ms
      Name: /var/lib/grafana/dashboards/ - Function: file.recurse - Result: Changed Started: - 20:45:56.887341 Duration: 732.704 ms
      Name: grafana-server - Function: service.running - Result: Changed Started: - 20:45:58.197514 Duration: 248.293 ms
      Name: /var/lib/grafana/dashboards - Function: file.directory - Result: Clean Started: - 20:45:58.447670 Duration: 0.971 ms

    Summary for salt-master
    -------------
    Succeeded: 57 (changed=43)
    Failed:     1
    -------------
    Total states run:     58
    Total run time:  288.786 s

No minions matched the target. No command was sent, no jid was assigned.
[ERROR   ] No minions returned
No minions matched the target. No command was sent, no jid was assigned.
[ERROR   ] No minions returned
No minions matched the target. No command was sent, no jid was assigned.
[ERROR   ] No minions returned
No minions matched the target. No command was sent, no jid was assigned.
[ERROR   ] No minions returned
salt-master_master:
  Name: set_up_consul_servers - Function: salt.state - Result: Changed Started: - 20:40:43.161691 Duration: 16388.534 ms
----------
          ID: set_up_consul_clients
    Function: salt.state
      Result: False
     Comment: No minions returned
     Started: 20:40:59.550657
    Duration: 43.543 ms
     Changes:
  Name: join_consul_members - Function: salt.state - Result: Changed Started: - 20:40:59.594354 Duration: 594.103 ms
----------
          ID: set_up_salt_master
    Function: salt.state
      Result: False
     Comment: Run failed on minions: salt-master
              Failures:
                  salt-master:
                    Name: unzip - Function: pkg.installed - Result: Clean Started: - 20:41:03.292658 Duration: 799.128 ms
                    Name: /opt/consul/agent - Function: archive.extracted - Result: Clean Started: - 20:41:04.092709 Duration: 681.057 ms
                    Name: /opt/consul/agent/consul - Function: file.managed - Result: Clean Started: - 20:41:04.773951 Duration: 1.322 ms
                    Name: /usr/bin/consul - Function: file.symlink - Result: Clean Started: - 20:41:04.776017 Duration: 0.891 ms
                    Name: /var/consul - Function: file.directory - Result: Clean Started: - 20:41:04.777043 Duration: 0.784 ms
                    Name: /etc/opt/consul.d - Function: file.directory - Result: Clean Started: - 20:41:04.777940 Duration: 0.794 ms
                    Name: /etc/opt/consul.d/ - Function: file.recurse - Result: Clean Started: - 20:41:04.778857 Duration: 146.589 ms
                    Name: dnsmasq - Function: pkg.installed - Result: Clean Started: - 20:41:04.925585 Duration: 0.533 ms
                    Name: /etc/dnsmasq.conf - Function: file.managed - Result: Changed Started: - 20:41:04.928108 Duration: 148.703 ms
                    Name: /etc/dnsmasq.d/10-consul - Function: file.append - Result: Changed Started: - 20:41:05.077020 Duration: 2.663 ms
                    Name: dnsmasq - Function: service.running - Result: Changed Started: - 20:41:05.080467 Duration: 106.88 ms
                    Name: /etc/resolv.conf - Function: file.prepend - Result: Changed Started: - 20:41:05.187554 Duration: 2.751 ms
                    Name: collectd - Function: pkg.installed - Result: Clean Started: - 20:41:05.190422 Duration: 0.507 ms
                    Name: collectd-postgresql.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:41:05.191059 Duration: 50408.015 ms
                    Name: collectd-java.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:41:55.711575 Duration: 31318.094 ms
                    Name: /lib64/libjvm.so - Function: file.symlink - Result: Changed Started: - 20:42:27.112308 Duration: 59.09 ms
                    Name: collectd-rrdtool.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:42:27.197114 Duration: 26954.658 ms
                    Name: collectd-generic-jmx.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:42:54.241479 Duration: 17436.368 ms
                    Name: collectd-write_sensu.x86_64 - Function: pkg.installed - Result: Changed Started: - 20:43:11.712289 Duration: 17791.937 ms
                    Name: /etc/collectd.conf - Function: file.managed - Result: Changed Started: - 20:43:30.168852 Duration: 266.859 ms
                    Name: /usr/share/collectd/types.d - Function: file.managed - Result: Changed Started: - 20:43:30.436396 Duration: 115.128 ms
                    Name: collectd - Function: service.running - Result: Changed Started: - 20:43:30.552648 Duration: 520.615 ms
                    Name: /var/log/collectd.log - Function: file.managed - Result: Changed Started: - 20:43:31.073546 Duration: 178.152 ms
                    Name: install_filebeat - Function: pkg.installed - Result: Changed Started: - 20:43:31.257289 Duration: 5533.408 ms
                    Name: /etc/filebeat/filebeat.yml - Function: file.managed - Result: Changed Started: - 20:43:36.793821 Duration: 243.455 ms
                    Name: filebeat - Function: service.enabled - Result: Changed Started: - 20:43:37.527077 Duration: 202.969 ms
                    Name: filebeat - Function: service.running - Result: Changed Started: - 20:43:37.730937 Duration: 69.706 ms
                    Name: libpcap - Function: pkg.installed - Result: Clean Started: - 20:43:37.805809 Duration: 0.473 ms
                    Name: install_packetbeat - Function: pkg.installed - Result: Changed Started: - 20:43:37.806381 Duration: 5844.761 ms
                    Name: /etc/packetbeat/packetbeat.yml - Function: file.managed - Result: Changed Started: - 20:43:43.677520 Duration: 218.495 ms
                    Name: packetbeat - Function: service.enabled - Result: Changed Started: - 20:43:44.387993 Duration: 160.621 ms
                    Name: packetbeat - Function: service.running - Result: Changed Started: - 20:43:44.548811 Duration: 87.587 ms
                    Name: ntp - Function: pkg.installed - Result: Changed Started: - 20:43:44.641448 Duration: 21328.646 ms
                    Name: ntpd - Function: service.enabled - Result: Changed Started: - 20:44:07.170765 Duration: 109.051 ms
                    Name: ntpd - Function: service.running - Result: Changed Started: - 20:44:07.280036 Duration: 174.655 ms
                    Name: /etc/opt/consul.d/server-config.json - Function: file.managed - Result: Clean Started: - 20:44:07.494604 Duration: 139.291 ms
                    Name: /usr/lib/systemd/system/consul-server.service - Function: file.managed - Result: Clean Started: - 20:44:07.634074 Duration: 56.638 ms
                    Name: /etc/opt/consul.d/influxdb_consul.json - Function: file.managed - Result: Changed Started: - 20:44:07.691808 Duration: 119.135 ms
                    Name: /etc/opt/consul.d/grafana_consul.json - Function: file.managed - Result: Changed Started: - 20:44:07.811161 Duration: 158.401 ms
                    Name: consul-server - Function: service.running - Result: Changed Started: - 20:44:08.008765 Duration: 2059.765 ms
                  ----------
                            ID: join-all-consul-members
                      Function: cmd.run
                          Name: consul join 10.0.0.59
                        Result: False
                       Comment: Command "consul join 10.0.0.59" run
                       Started: 20:44:10.221437
                      Duration: 69.428 ms
                       Changes:
                                ----------
                                pid:
                                    14625
                                retcode:
                                    1
                                stderr:
                                    Error joining address '10.0.0.59': Put http://127.0.0.1:8500/v1/agent/join/10.0.0.59: dial tcp 127.0.0.1:8500: getsockopt: connection refused
                                    Failed to join any nodes.
                                stdout:
                    Name: influxdb - Function: pkg.installed - Result: Changed Started: - 20:44:10.327802 Duration: 31209.716 ms
                    Name: influxdb - Function: user.present - Result: Changed Started: - 20:44:42.446100 Duration: 282.127 ms
                    Name: /etc/influxdb/influxdb.conf - Function: file.managed - Result: Changed Started: - 20:44:42.730048 Duration: 147.816 ms
                    Name: influxdb - Function: service.running - Result: Changed Started: - 20:44:42.879108 Duration: 84.587 ms
                    Name: /var/lib/.influxdb - Function: file.directory - Result: Changed Started: - 20:44:42.963972 Duration: 1.546 ms
                    Name: /var/lib/.influxdb/data - Function: file.directory - Result: Changed Started: - 20:44:42.965636 Duration: 1.256 ms
                    Name: /var/lib/.influxdb/meta - Function: file.directory - Result: Changed Started: - 20:44:42.967029 Duration: 1.298 ms
                    Name: /var/lib/.influxdb/hh - Function: file.directory - Result: Changed Started: - 20:44:42.968449 Duration: 1.251 ms
                    Name: /var/lib/.influxdb/wal - Function: file.directory - Result: Changed Started: - 20:44:42.969842 Duration: 1.283 ms
                    Name: initscripts - Function: pkg.installed - Result: Clean Started: - 20:44:42.988186 Duration: 0.517 ms
                    Name: fontconfig - Function: pkg.installed - Result: Clean Started: - 20:44:42.988815 Duration: 0.467 ms
                    Name: grafana - Function: pkgrepo.managed - Result: Changed Started: - 20:44:43.006339 Duration: 519.696 ms
                    Name: grafana - Function: pkg.installed - Result: Changed Started: - 20:44:43.526208 Duration: 71889.838 ms
                    Name: /etc/grafana/grafana.ini - Function: file.managed - Result: Changed Started: - 20:45:56.742311 Duration: 144.832 ms
                    Name: /var/lib/grafana/dashboards/ - Function: file.recurse - Result: Changed Started: - 20:45:56.887341 Duration: 732.704 ms
                    Name: grafana-server - Function: service.running - Result: Changed Started: - 20:45:58.197514 Duration: 248.293 ms
                    Name: /var/lib/grafana/dashboards - Function: file.directory - Result: Clean Started: - 20:45:58.447670 Duration: 0.971 ms

                  Summary for salt-master
                  -------------
                  Succeeded: 57 (changed=43)
                  Failed:     1
                  -------------
                  Total states run:     58
                  Total run time:  288.786 s
     Started: 20:41:00.188624
    Duration: 298533.057 ms
     Changes:
----------
          ID: set_up_db_server
    Function: salt.state
      Result: False
     Comment: No minions returned
     Started: 20:45:58.721851
    Duration: 66.965 ms
     Changes:
----------
          ID: set_up_job_queue
    Function: salt.state
      Result: False
     Comment: No minions returned
     Started: 20:45:58.788989
    Duration: 42.914 ms
     Changes:
----------
          ID: set_up_tracker
    Function: salt.state
      Result: False
     Comment: No minions returned
     Started: 20:45:58.832079
    Duration: 39.181 ms
     Changes:
  Name: set_up_monitoring_server - Function: salt.state - Result: Changed Started: - 20:45:58.871422 Duration: 18945.863 ms
----------
          ID: set_up_workers
    Function: salt.state
      Result: False
     Comment: No minions returned
     Started: 20:46:17.817456
    Duration: 50.162 ms
     Changes:

Summary for salt-master_master
------------
Succeeded: 3 (changed=3)
Failed:    6
------------
Total states run:     9
Total run time: 334.704 s
```

Do I have to spawn other VMs manually after the terraform succeeds? Modify the terraform definitions to spawn some workers/dbs/etc...?

Here's what consul sees so far:

```
[root@salt-master centos]# consul members
Node         Address         Status  Type    Build  Protocol  DC
salt-master  10.0.0.59:8301  alive   server  0.8.1  2         butler_dc
```

And only two t2.micro instances are spawned on the EC2 console: butler-jump and saltmaster.

**Comment by yakneens** [comment]
I will look into this.

**Comment by yakneens** [comment]
Looks like some files were missed when this was initially committed. I will do some testing to make sure everything is still fine and push those out.

**Comment by yakneens** [comment]
@brainstorm, FYI, I have committed the missing files (https://github.com/llevar/butler/commit/e44967012823beae0cfab334f400709c29368c3f), however, after a recent upgrade of Consul there seems to be a regression on AWS, whereby the orchestrated deployment is no longer stable due to service name resolution delays, i.e. some service names are unresolvable at the time that they are requested by the deployment procedure. It's still possible to deploy (and I validated this successfully) by issuing the same commands the orchestrator is running manually (https://github.com/llevar/butler/blob/master/configuration/salt/state/butler/deploy/init.sls). 

Basically run these on the salt-master:

salt '\*' state.apply consul
salt-call state.apply consul.join-all
consul members **** This should print out all the nodes ***
salt salt-master state.highstate
salt db-server state.highstate
salt job-queue state.highstate
salt tracker state.highstate
salt worker* state.highstate

This should get everything deployed. In the meantime I will try to figure out what is causing the name resolution to be delayed during the automated deployment process. This issue does not seem to affect OpenStack at this time.

Thanks for your patience.

**Comment by ohofmann** [comment]
Sergei, just a quick 'thanks!' from me as well. Hope to catch up soon.

**Comment by yakneens** [comment]
@brainstorm, @ohofmann , Ok, after a bit of massaging AWS is back in green territory. Give it a whirl. Keep in mind the deployment flavours - right now this deploys 5 m4.xlarge instances. This is what's needed for real payloads. You may be able to downsize if you tweak some settings, like number of database connections. Eventually I will release a small-cluster analog of the deployment scripts.

**Comment by brainstorm** [comment]
Thanks a lot @llevar for digging into this and for the fix, much appreciated :)

I reckon that just changing `vars.tf` so that it instantiates smaller (and perhaps spot instances) is all what's needed to spawn a small test cluster?

I hope you have the chance to speak with @ohofmann soon, happy to give it another spin if time allows ;)

**Comment by yakneens** [comment]
@brainstorm, yea, just replace the flavour. I haven't experimented with spot market in the Butler context but Terraform (the underlying provisioning framework) has rich support for AWS spot market. I would not recommend launching any management nodes on the spot market though, only the workers.
